sstables/sstables.cc

```diff
@@ -1396,6 +1399,31 @@ void sstable::write_filter() {
     write_simple<component_type::Filter>(filter_ref);
 }
 
+size_t sstable::total_reclaimable_memory_size() const {
+    if (!_total_reclaimable_memory) {
+        _total_reclaimable_memory = _components->filter ? _components->filter->memory_size() : 0;
+    }
+
+    return _total_reclaimable_memory.value();
+}
+
+size_t sstable::reclaim_memory_from_components() {
+    size_t total_memory_reclaimed = 0;
+
+    if (_components->filter) {
+        auto filter_memory_size = _components->filter->memory_size();
+        if (filter_memory_size > 0) {
+            // discard it from memory by replacing it with an always present variant
+            _components->filter = std::make_unique<utils::filter::always_present_filter>();
+            _recognized_components.erase(component_type::Filter);
+            total_memory_reclaimed += filter_memory_size;
+        }
```

Comment:
I lack code context, so it's faster for me to ask (ignore or close if it's not possible).
Is it possible the big bloom filter will be reloaded again by some other component/code  (streaming? repair?)
and we end up in a reclaim loop?

I am asking because bloom filter on disk can change only with compaction, so I'd like to understand if it can be loaded back and when.

Reply:
Hi @tarzanek! 

- For existing SSTables, the bloom filters are read from disk only when a node starts and all the data directories are processed to load the SSTables' metadata into memory.
- For new SSTables that are created by flushing memtables, the bloom filter gets created in-memory. The filters will be written to disk but as we already have the filter in memory, a reload back from disk never happens.

So, there is no chance of ending up in a reclaim loop.

In both the cases, we replace it with a `always_present_filter` variant, and the modules using this filter will not see any difference except getting a `true` for all `is_present(key)` requests and as a result going into every SSTable for reading.

Right now there is no way to reload the reclaimed bloom filters but there is a proposed follow-up to this PR that will reload these reclaimed bloom filters when memory becomes available eventually.